### ISSUE TITLE
Train not being generated?

### ISSUE THREAD

**Issue by eliot-akira** [SimHacker/MicropolisCore#4]
With the JavaScript port MicropolisJS, I remember I found and fixed a bug in the scanning (?) logic that was preventing trains from being generated.

It looks like the same bug exists in MicropolisEngine - or rather, the port faithfully translated the bug into JS. As far as I can see there are no trains running in any of the cities.

I forgot specifically which part of the code was causing the issue, but it was something small, like a single conditional statement. Debugging the C++ code is somewhat challenging (well for one I'm not so familiar with the language) - I think I'll temporarily put a JS callback that receives values and outputs to console log.

Anyway, I'll keep looking, tracing the logic - will make a pull request if I solve it.

**Comment by SimHacker** [comment]
Good catch! I’ll dig through the old code to see how it generates trains. Could you please open an issue so it won’t fall through the cracks if I forget? 

I’m re-reading the Snap! Manual and extensions documentation and code, to figure out how to integrate it with Micropolis. 

Right now it’s just putting a big draggable snap window over the Micropolis view which is full screen, but they need to be the other way around, with the Micropolis view embedded in the Snap! stage.

From reading the code it turns out there is a way to do that, which the video camera and the map features use, that give Snap a canvas to draw as the background that constantly updates. 

That means the Micropolis view will have to be off screen, and Snap will implement the mouse tracking. 

But I still want to be able to run Micropolis without Snap of course, so I’m refactoring the big blob of code in MicropolisView.svelte into independent TypeScript class and svelte components like MicropolisSimulator.ts, MicropolisView.svelte, MicropolisCallbackLog.ts, so we can make different implementations for MicropolisViewSnap.svelte and MicropolisCallbackSnap.ts. 

-Don

**Comment by eliot-akira** [comment]
I went spelunking into the codebase, added a method `Micropolis::log()` to send a string to JS console log, and traced the logic of railroads and trains. It was interesting to inspect values in real time as the engine went through its cycles. A log method like this seems handy for sending arbitrary values, possibly even JSON string.

As far as I could tell, the train logic is working correctly. It's just that the JS side is not rendering any sprites yet. So there's no train, helicopter, boat, etc.

So I'll change the title of this issue, as a reminder about sprite render.

**Comment by eliot-akira** [comment]
That makes sense about putting the Micropolis canvas into Snap, and not the other way around. It sounds like it may involve an overhaul of the web interface, to use Snap as the foundation.

By the way, I was able to extract the TileRenderer and use it independently, including mouse navigation, in a new empty project with a static page. During the process, I found that `WebGLTileRenderer` works but not `CanvasTileRenderer` with the same interface, maybe due to some change in the data schema. I like how the canvas rendering logic is much simpler than WebGL, so when I have time I plan to study it deeper and see if I can get it working.

**Comment by SimHacker** [comment]
Snap has some hooks to enable drawing a real time animated stage background, for video and maps, that I think would work. 

https://github.com/jmoenig/Snap/blob/master/src/objects.js#L8638

    // projection layer - for video, maps, 3D extensions etc., transient
    this.projectionSource = null; // offscreen DOM element for video, maps, 3D
    this.getProjectionImage = null; // function to return a blittable image
    this.stopProjectionSource = null; // function to turn off video stream etc.
    this.continuousProjection = false; // turn ON for video
    this.projectionCanvas = null;
    this.projectionTransparency = 50;

https://github.com/jmoenig/Snap/blob/master/src/objects.js#L8774

    // projection layer (e.g. webcam)
    if (this.projectionSource) {
        ctx.globalAlpha = 1 - (this.projectionTransparency / 100);
        ctx.drawImage(
            this.projectionLayer(),
            sl / this.scale,
            st / this.scale,
            ws,
            hs,
            clipped.left() / this.scale,
            clipped.top() / this.scale,
            ws,
            hs
        );
        this.version = Date.now(); // update watcher icons
    }

https://github.com/jmoenig/Snap/blob/master/src/objects.js#L8864

StageMorph.prototype.startVideo = function() {
    var myself = this;

    function noCameraSupport() {
        var dialog = new DialogBoxMorph();
        dialog.inform(
            localize('Camera not supported'),
            localize('Please make sure your web browser is up to date\n' +
                'and your camera is properly configured. \n\n' +
                'Some browsers also require you to access Snap!\n' +
                'through HTTPS to use the camera.\n\n' +
                'Please replace the "http://" part of the address\n' +
                'in your browser by "https://" and try again.'),
            this.world
        );
        dialog.fixLayout();
        if (myself.projectionSource) {
            myself.projectionSource.remove();
            myself.projectionSource = null;
        }
    }
    if (this.projectionSource) { // video capture has already been started
        return;
    }

    this.projectionSource = document.createElement('video');
    this.projectionSource.width = this.dimensions.x;
    this.projectionSource.height = this.dimensions.y;
    this.projectionSource.hidden = true;
    document.body.appendChild(this.projectionSource);
    if (!this.videoMotion) {
        this.videoMotion = new VideoMotion(
            this.dimensions.x,
            this.dimensions.y
        );
    }
    if (navigator.mediaDevices && navigator.mediaDevices.getUserMedia) {
        navigator.mediaDevices.getUserMedia({ video: true })
            .then(function(stream) {
                myself.getProjectionImage = myself.getVideoImage;
                myself.stopProjectionSource = myself.stopVideo;
                myself.continuousProjection = true;
                myself.projectionSource.srcObject = stream;
                myself.projectionSource.play().catch(noCameraSupport);
                myself.projectionSource.stream = stream;
            })
            .catch(noCameraSupport);
    }
};

StageMorph.prototype.getVideoImage = function () {
    return this.projectionSource;
};

StageMorph.prototype.stopVideo = function() {
    if (this.projectionSource && this.projectionSource.stream) {
        this.projectionSource.stream.getTracks().forEach(track =>
            track.stop()
        );
    }
    this.videoMotion = null;
};

StageMorph.prototype.stopProjection = function () {
    if (this.projectionSource) {
        this.stopProjectionSource();
        this.projectionSource.remove();
        this.projectionSource = null;
        this.continuousProjection = false;
    }
    this.clearProjectionLayer();
};

StageMorph.prototype.projectionSnap = function (target) {
    var snap = newCanvas(this.dimensions, true),
        ctx = snap.getContext('2d');
    ctx.drawImage(this.projectionLayer(), 0, 0);
    return new Costume(snap, (target || this).newCostumeName(localize('snap')));
};

**Comment by SimHacker** [comment]
And of course it should be general enough to support the SimCity sprite animation (and other kinds of sprites), as additional layers. 

Now that I think of it, one thing that WebGPU might give us (but I don't know for sure) is the ability to pass in more data structures and texture layers than OpenGL supports. They both have limits, but in general WebGPU is a hell of a lot more flexible, you don't have to stuff data structures into textures and hacks like that.

**Comment by SimHacker** [comment]
Yeah debugging the C++ code compiled into WebAssembly is practically impossible, unless you like to look at raw WebAssembly instructions, in which case it's so much fun! ;) So for people without so much time on their hands, printf debugging is the way to go, since it recompiles in only a few seconds! 

I still have to figure out how to configure Visual Studio Code to automatically recompile the C++ code and hot-redeploy the web app whenever I change a line of C++, but maybe it's better to have a little buffering in there since C++ is such a big pain in the ass that I'd hate to break the running app every time I change one character. 

The WebAssembly runtime has a stdout/stderr reporting interface that you can hook into in JavaScript when you make the web module. In the latest overhaul where I factored out MicropolisSimulator and TileView (in preparation for integrating with Snap!), I defined some stdout and stderr hooks that prefix the string with the name of the module so you know where it came from. 

https://github.com/SimHacker/MicropolisCore/blob/main/micropolis/src/lib/MicropolisSimulator.ts#L14

    micropolisengine = {
        print: (message: string) => console.log("micropolisengine:", message),
        printErr: (message: string) => console.error("micropolisengine: ERROR: ", message),
        setStatus: (status: string) => console.log("micropolisengine: initModule: status:", status),
        locateFile: (path: string, prefix: string) => {
          console.log("micropolisengine: initModule: locateFile:", "prefix:", prefix, "path:", path);
          return prefix + path;
        },
        onRuntimeInitialized: () => console.log("micropolisengine: onRuntimeInitialized:"),
      };

      await initModule(micropolisengine);

**Comment by SimHacker** [comment]
Oh as far as inspecting the internal state of the simulator, I want to make an optional way to dump that ALL out as JSON, including the 2d map overlay layers (even the whole save file, so you can rewind and replay from any point), so it can send a stream of telemetry to a Promethius time series database at regular intervals (or every step for debugging), so that we can make Grafana dashboard and panels that query and visualize it -- that will be GREAT for debugging, as well as playing the game (giving much deeper insights than just the traditional history and evaluation dialogs and RCI gauge do). 

https://prometheus.io/

Rather than implementing this feature in C++ (since C++ doesn't have good JSON support), I think it's best to just systematically expose EVERYTHING about the simulator state to JavaScript (only ignoring a few things that really don't matter), and then writing the code to pull it out into a JSON+binary blob, then stream that to the time series database, all in JavaScript. That would serve as a good test that all the simulator state was exposed correctly. There are still a bunch of things missing, and the emscripten.cpp wrapper file needs to be better organizes so it's easier to tell what's been wrapped and what hasn't.

**Comment by SimHacker** [comment]
It would also be cool to get the WASM simulator running in the node server, so the server can have an endpoint that will return a safe file at any point in time, by reconstructing it from the most recent save file plus all simulation steps and edits. 

There are a lot of other reasons to run the simulator on the web server too (like the sheer beauty of the symmetry), but that's one example. The Python version of Micropolis ran multiple instances of the C++ simulation on the Python web server, and any number of web browser / Flash clients could connect to them. But the technology became obsolete before I had a chance to finish a full blown multi player interface. But now that the simulator runs just fine in the client, a better approach is running it in synchronized lockstep in all the clients, just bouncing editing and chatting and content sharing messages off of the server. But running the simulation on the server might be useful for video streaming and mobile interfaces, and stuff like that.

**Comment by SimHacker** [comment]
Update (Jun 2026): Confirmed still accurate — engine simulates sprites (SimSprite); TileView does not composite them yet.

2024: @eliot-akira ruled out train-generation bug; renamed to rendering.
2026: Per-set sprite PNGs preserved colocated with renderer atlases (-sprite-.png, e.g. asia-sprite-chopper.png) — commit https://github.com/SimHacker/MicropolisCore/commit/5c428ebd6bbff4912d48d4dbe0a5a69575cd6ac3. Source BMPs remain in content/micropolis/tilesets//.
Implementation tracked in repo design docs (not a new issue): map-compositing §2.4 sprites compositor, §2.5 per-tileset packs, documentation/TODO.md R1.
Remaining work: wire engine sprite snapshots → compositor (interactive WebGPU/holodeck + software path for print/overview), resolving atlas from active tileset pack. Mono/snow sprites still TBD (Mac resource extraction).

**Comment by SimHacker** [comment]
## Rendering roadmap (follow-up — Jun 2026)

Building on the update above: this issue stays open as the tracker for **sprite compositing**. The engine side is fine; the gap is entirely on the **client compositor**.

---

### What works today

| Layer | Status | Where |
|-------|--------|-------|
| **Simulation** | Trains, helicopters, ships, monsters, tornadoes, explosions simulated | C++ / WASM — `SimSprite` in [`packages/micropolis-engine`](https://github.com/SimHacker/MicropolisCore/tree/main/packages/micropolis-engine) |
| **Tile map** | Pan, zoom, stacked atlases (`all.png`), per-cell tile lookup | [`packages/tile-renderer/src/WebGPUTileRenderer.ts`](https://github.com/SimHacker/MicropolisCore/blob/main/packages/tile-renderer/src/WebGPUTileRenderer.ts) via [`createMapTileRenderer()`](https://github.com/SimHacker/MicropolisCore/blob/main/packages/tile-renderer/src/createMapTileRenderer.ts) — default chain **WebGPU → canvas (software)** |
| **Sprite metadata** | Frame sizes, hotspots, directions per agent type | [`apps/micropolis/src/lib/sprites.ts`](https://github.com/SimHacker/MicropolisCore/blob/main/apps/micropolis/src/lib/sprites.ts) |
| **Sprite art (preserved)** | Per-tileset PNG sheets colocated with atlases | `apps/micropolis/src/lib/images/tilesets/` — e.g. `asia-sprite-chopper.png`, `classic-sprite-monster.png` ([`5c428eb`](https://github.com/SimHacker/MicropolisCore/commit/5c428eb)); source BMPs in [`content/micropolis/tilesets/`](https://github.com/SimHacker/MicropolisCore/tree/main/content/micropolis/tilesets) |
| **Sprites on screen** | **Not implemented** | `TileView.svelte` draws tiles only — no compositor pass yet |

@eliot-akira's 2024 finding still holds: you can trace railroad/train logic in the engine and see sprites being created; they just never get blitted in the browser.

---

### Where this fits in the architecture

We're converging on a **Holodeck WebGPU compositor** — one display-list executor where map tiles, sprites, cursors, pie menus, and data overlays are **plugin layers**, not ad-hoc draws inside `TileView`:

- [**Unified WebGPU renderer**](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/unified-webgpu-renderer.md) — holodeck layers, pick buffer, Micropolis sprite as layer type 8
- [**Renderer plugin roadmap**](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/renderer-plugin-roadmap.md) — ordered TODO: holodeck executor → `MicropolisMapPlugin` → overlays → **tileset packs + sprites** → migrate `TileView` → `HolodeckStage`
- [**Map compositing & measurement**](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/map-compositing-and-measurement.md) — dual-path rule: **software (Node/print/CI) and WebGPU (interactive) must match**

**Gating:** holodeck migration and sprite plugins land **after playable A–C** (HUD, tools, click-to-build, classic feedback) — see [playable game readiness](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/micropolis-playable-game-readiness.md). The current WebGL/WebGPU tile bridge in `TileView` stays the map rasterizer for the vertical slice until then.

---

### Sprite compositor spec (authoritative)

[**Map compositing §2.4 — Sprites**](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/map-compositing-and-measurement.md#24-sprites--required-on-software-print-iconic-maps-overviews)

- **Inputs:** engine `SimSprite` snapshot (type, world position, frame) + [`sprites.ts`](https://github.com/SimHacker/MicropolisCore/blob/main/apps/micropolis/src/lib/sprites.ts) hotspot/frame metadata + **active tileset pack** sprite sheet
- **Interactive client:** animated blit every frame via WebGPU `MicropolisSpritePlugin` (holodeck)
- **Software path (required, not optional):** server `/render`, print, thumbnails, OG images — same data, same `SpriteRenderProfile`, including **overview mode** (scale-around-hotspot + drop shadow so trains/monsters read at city-wide zoom)
- **Parity test:** software vs WebGPU pixel diff on fixed fixtures

[**Map compositing §2.5 — Tileset packs & per-set sprites**](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/designs/map-compositing-and-measurement.md#25-tileset-packs-mop-mixing-and-per-set-sprites)

Helicopter / train / monster art is **not global** — Ancient Asia `chopper.bmp` ≠ Classic. Compositor resolves:

`SimSprite.type` + `frame` + **active tileset pack** → rect in the correct `-sprite-*.png` sheet (or plugin override).

Seven color sets preserved today; mono/snow sprite sheets still TBD (Mac resource extraction).

---

### Backends (2026 vs 2024 thread above)

A lot changed since the Jun 2024 comments about WebGL vs Canvas:

| Backend | Role today |
|---------|------------|
| **WebGPU** | Default interactive map ([`WebGPUTileRenderer`](https://github.com/SimHacker/MicropolisCore/blob/main/packages/tile-renderer/src/WebGPUTileRenderer.ts)) |
| **Canvas / software** | Browser fallback + Node `/render` + CI — pixel loop, no GPU ([`packages/render-core/src/raster/software.ts`](https://github.com/SimHacker/MicropolisCore/blob/main/packages/render-core/src/raster/software.ts)) |
| **WebGL** | **Frozen / opt-in only** — not in default chain; legacy bridge until holodeck cutover |

Sprites will follow the same split: **WebGPU plugin for the tab, software compositor for export** — not a separate Canvas per-sprite blit path.

Classic Micropolis tiles are **pre-rendered iso bitmaps** — orientation is baked into atlas indices, not quad rotation. (The renderer's `tileRotate` uniform is a dev-only atlas index offset for the heat/CA demo, not map rotation.)

Future **3D city rendering** (Blender assets, etc.) is a separate lower-priority track — this issue is about the retro sprite sheets.

---

### Implementation order (from [TODO.md R1](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/TODO.md#micropolis-renderer))

1. Playable vertical slice A–C on existing `TileView` map
2. `HolodeckStage` executor + `MicropolisMapPlugin` (absorb tile renderer)
3. **Software sprite compositor** + WebGPU `MicropolisSpritePlugin` ([#4](https://github.com/SimHacker/MicropolisCore/issues/4))
4. MOP overlay layers (police/fire coverage, pollution, traffic — community interest from [#8](https://github.com/SimHacker/MicropolisCore/pull/8) too)
5. Full `TileView` → `HolodeckStage` cutover

Work index: [`documentation/TODO.md`](https://github.com/SimHacker/MicropolisCore/blob/main/documentation/TODO.md)

---

### Related closed / merged work

- [#9](https://github.com/SimHacker/MicropolisCore/issues/9) — Ancient Asia tile atlas fixed (`asia.bmp`)
- [#5](https://github.com/SimHacker/MicropolisCore/issues/5) — tile edge duplication; closed (not reproduced on WebGPU → canvas default)
- [#8](https://github.com/SimHacker/MicropolisCore/pull/8) — @DunDarach WebGPU experiments; closed in favor of monorepo + holodeck path
- [PR #6](https://github.com/SimHacker/MicropolisCore/pull/6) — concurrent Vite + C++ watch; reimplemented on monorepo ([`94524ff`](https://github.com/SimHacker/MicropolisCore/commit/94524ff))

---

**Bottom line:** keep this issue open until something in the repo actually composites `SimSprite` snapshots — first software (print/export), then WebGPU holodeck parity. Asset preservation is done; wiring is not.
